include/jemalloc/internal/extent_structs.h

```diff
@@ -102,8 +102,20 @@ struct extent_s {
 	/* Pointer to the extent that this structure is responsible for. */
 	void			*e_addr;
 
-	/* Extent size. */
-	size_t			e_size;
+	union {
+		/*
+		 * Extent size and serial number associated with the extent
+		 * structure (different than the serial number for the extent at
+		 * e_addr).
+		 *
+		 * ssssssss [...] ssssssss ssssnnnn nnnnnnnn
+		 */
+		size_t			e_size_esn;
```

Comment:
The serial number is the primary comparison key, and the address is secondary.  Before the introduction of serial numbers, we used address as the primary (and only) key, but if memory grew downward (as on Linux), then address-only ordering caused us to prefer newer mappings, which subverted the packing heuristic.  The serial number allows us to prefer old mappings, and therefore consistently pack.

The above describes the general situation that motivates serial numbers.  However, we need a separate serial number for the extent (sn for virtual memory that actually backs application allocations) versus the extent structure serial number (esn for metadata which came from the base allocator).